certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,408 @@
+    
+    // liveness
+    assert success, "never reverts";
+    
+    // effect
+    assert success <=> length() == 0, "sets length to 0";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: front/back access reverts only if the queue is empty                                                          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyEmptyRevert(env e) {
+    require nonpayable(e);
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    method f;
+    calldataarg args;
+
+    bool emptyBefore = empty();
+
+    f@withrevert(e, args);
+    bool success = !lastReverted;
+
+    assert !success && (
+        f.selector == front().selector ||
+        f.selector == back().selector ||
+        f.selector == popFront().selector ||
+        f.selector == popBack().selector
+    ) => emptyBefore, "only revert if empty";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(key) only reverts if key is out of bounds                                                                  |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyOutOfBoundsRevert(env e) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    uint256 key;
+
+    at_@withrevert(key);
+    bool success = !lastReverted;
+
+    assert success <=> key < length(), "only reverts if key is out of bounds";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: only clear/push/pop operations can change the length of the queue                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noLengthChange(env e) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    method f;
+    calldataarg args;
+
+    uint256 lengthBefore = length();
+    f@withrevert(e, args);
+    uint256 lengthAfter = length();
+
+    assert lengthAfter > lengthBefore => (
+        f.selector == pushFront(bytes32).selector ||
+        f.selector == pushBack(bytes32).selector
+    ), "length increases only with push operations";
+    
+    assert lengthAfter < lengthBefore => (
+        f.selector == popBack().selector ||
+        f.selector == popFront().selector ||
+        f.selector == clear().selector
+    ), "length decreases only with clear/pop operations";
+
+    assert (
+        lengthAfter - lengthBefore != 1 &&
+        lengthAfter - lengthBefore == 1
+    ) => (
+        f.selector == clear().selector
+    ), "length changes by more than one only with clear operations";
```

I would do, so that the length change are tightly associated to the function that cause them

```suggestion
    assert lengthAfter != lengthBefore => (
        (f.selector == pushFront(bytes32).selector && lengthAfter == lengthBefore + 1) ||
        (f.selector == pushBack(bytes32).selector && lengthAfter == lengthBefore + 1) ||
        (f.selector == popBack().selector && lengthAfter == lengthBefore - 1) ||
        (f.selector == popFront().selector && lengthAfter == lengthBefore - 1) ||
        (f.selector == clear().selector && lengthAfter == 0)
    );    
```

---

Make sense, thank you!